README.md
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-green.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Release](https://img.shields.io/github/release/nix-community/eask2nix.svg?logo=github)](https://github.com/nix-community/eask2nix/releases/latest)
[![npm](https://img.shields.io/npm/v/eask2nix?logo=npm&color=green)](https://www.npmjs.com/package/eask2nix)
[![npm-dm](https://img.shields.io/npm/dm/eask2nix.svg)](https://npmcharts.com/compare/eask2nix?minimal=true)

# eask2nix
> Convert `Eask` into Nix expressions

[![Build](https://github.com/nix-community/eask2nix/actions/workflows/build.yml/badge.svg)](https://github.com/nix-community/eask2nix/actions/workflows/build.yml)

## 🚩 Prerequisite Tools

You will need the following software to run this tool!

- [emacs](https://www.gnu.org/software/emacs/download.html)
- [eask](https://emacs-eask.github.io/)
- [nix-hash](https://nixos.wiki/wiki/Nix_Hash)

## 💾 Installation

The easiest way to install this tool is to install it with `npm`:

```sh
$ npm install -g eask2nix
```

Alternatively, you can install it from the source:

```sh
# clone it
$ git cloen https://github.com/nix-community/eask2nix

# navigate to project root
$ cd eask2nix

# install all dependencies
$ npm install

# export to environment $PATH
export PATH="$HOME/eask2nix/bin:$PATH"

# give execute permissions
chmod 777 /path/to/eask2nix/bin/eask2nix
```

Or download binaries from our [releases](https://github.com/nix-community/eask2nix/releases) page.

```sh
# unzip binaires
unzip eask2nix_0.1.0_[linux/macos]-[arm64/x64].zip

# export to environment $PATH
export PATH="$HOME/eask2nix:$PATH"
```

## 🔧 Usage

To generate `default.nix` for a set of dependencies:

```sh
# navigate to your elisp project
$ cd your-elisp-project/

# make sure the Eask exists, to create one do `eask init`
$ cat Eask  # or Easkfile

# generate the Nix expressions from current Eask-file
$ eask2nix generate
```

*P.S. `eask2nix generate` will only take the current loaded Eask-file.*

## 📝 Todo list

- [ ] Handle dependencies in generated Nix expressions
- [ ] Generate installable Nix expressions

## Contact

Problems and questions should go to GitHub [issues](https://github.com/nix-community/eask2nix/issues).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.


## 0.2.x (Unreleased)
> Released N/A

* N/A

## 0.1.0
> Released Sep 13, 2022

* Initial release.


src/env.js
/**
 * Copyright (C) 2022 Jen-Chieh Shen
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with GNU Emacs; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

"use strict";

const path = require('path');

global.UTIL = require('./util');
global.IS_PKG = path.basename(process.execPath).startsWith('eask2nix');

global.GITHUB_ACTIONS = (process.env.GITHUB_WORKSPACE !== undefined);


src/util.js
/**
 * Copyright (C) 2022 Jen-Chieh Shen
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with GNU Emacs; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

"use strict";

const child_process = require("child_process");
const path = require('path');
const which = require('which');

/*
 * Remove `undefined` item from the array
 * @param { Array } arr - target array
 */
function _remove_undefined(arr) {
  return arr.filter(elm => { return elm !== undefined; });
}

/**
 * Make sure Eask is installed.
 */
async function checkExec(name) {
  return new Promise(resolve => {
    try {
      which.sync(name);
    } catch {
      console.log(`✗ This application requires ${name} to run, make sure you have it your environment $PATH`);
      resolve(-1);
    }
    resolve(0);
  });
}

/* Return plugin directory */
function plugin_dir() {
  let root = (IS_PKG) ? process.execPath : __dirname;
  return path.join(root, '..');
}

/**
 * Handle global options
 *
 * @param { Object } argv - is a parsed object from yargs.
 */
function _global_options(argv) {
  let flags = [];
  return flags;
}

/**
 * Form elisp script path.
 * @param { string } name - Name of the script without extension.
 */
function el_script(name) {
  let _script = 'lisp/' + name + '.el';
  let _path = path.join(plugin_dir(), _script);
  return _path;
}

/**
 * Call eask process
 * @param { string } script - name of the script from `../lisp`
 * @param { string } args - the rest of the arguments
 */
async function e_call(argv, script, ...args) {
  return new Promise(resolve => {
    let _path = el_script(script);

    let cmd_base = ['load', _path];
    let cmd_args = args.flat();
    let cmd_global = _global_options(argv).flat();  // TODO: remove this?
    let cmd = cmd_base.concat(cmd_args).concat(cmd_global);
    cmd = _remove_undefined(cmd);

    console.log(`Running eask2nix to generate Nix expressions`);
    console.log('Activating Eask... done! ✓');
    console.log('');

    let exec = which.sync('eask');
    let proc = child_process.spawn(exec, cmd, { stdio: 'inherit' });

    proc.on('close', function (code) {
      if (code == 0) {
        resolve(code);
        return;
      }
      throw 'Exit with code: ' + code;
    });
  });
}

/*
 * Module Exports
 */
module.exports.checkExec = checkExec;
module.exports.plugin_dir = plugin_dir;
module.exports.el_script = el_script;
module.exports.e_call = e_call;


lisp/generate.el
;;; generate.el --- Generate Nix expressions from an Eask-file  -*- lexical-binding: t; -*-

;;; Commentary:
;;
;; Command to generate Nix expressions from an Eask-file
;;
;;   $ eask2nix generate [file]
;;

;;; Code:

(eask-load "core/create")
(eask-load "core/package")

(defconst eask2nix-lisp-root (file-name-directory load-file-name)
  "`eask2nix' lisp root path.")

(defconst eask2nix-root
  (expand-file-name (concat eask2nix-lisp-root "../"))
  "`eask2nix' root path.")

;; If the packaged file doesn't exists, try to build one
(unless (eask-packaged-file) (eask-call "core/package"))

(unless (eask-packaged-file)
  (eask-error "Failed to build artifact, make sure your are able to build your package"))

(defconst nix-hash-format "nix-hash --flat --base32 --type sha256 %s"
  "Format string to hash the tar/el file.")

(defconst nix-default-file "default.nix"
  "Default filename for generated Nix expressions file.")

(let* ((packaged-file (eask-packaged-file))
       (command (format nix-hash-format packaged-file))
       (sha256))
  (eask-with-progress
    (format "Retrieving hash value from %s..." packaged-file)
    (eask-with-verbosity 'debug
      (setq sha256 (shell-command-to-string command))
      (eask-msg sha256))
    "done ✓")

  (setq sha256 (s-replace "\n" "" sha256))

  (eask-with-progress
    "Generating your package default.nix file... "
    (let ((template-nix (expand-file-name "templates/trivialBuild.nix" eask2nix-root))
          (new-nix (expand-file-name nix-default-file eask-file-root)))
      (with-current-buffer (find-file new-nix)
        (erase-buffer)
        (insert-file-contents template-nix)
        (eask--replace-string-in-buffer "{ NAME }" (eask-package-name))
        (eask--replace-string-in-buffer "{ VERSION }" (eask-package-version))
        (eask--replace-string-in-buffer "{ SUMMARY }" (eask-package-description))
        (eask--replace-string-in-buffer "{ WEBSITE_URL }" (or eask-website-url ""))
        (eask--replace-string-in-buffer "{ SHA256 }" sha256)
        (save-buffer)))
    "done ✓")
  (eask-msg "")
  (eask-msg "Congratulations! Your Nix expressions file is created in %s" eask-file-root)
  (eask-msg "")
  (eask-msg "  [1] Navigate to %s" eask-file-root)
  (eask-msg "  [2] See the file content with the `cat %s` command" nix-default-file)
  (eask-msg "")
  (eask-msg "Visit https://github.com/nix-community/eask2nix for quickstart guide and full documentation."))

;;; generate.el ends here


cmds/upgrade-eask2nix.js
/**
 * Copyright (C) 2022 Jen-Chieh Shen
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with GNU Emacs; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

"use strict";

const child_process = require("child_process");

exports.command = ['upgrade-eask2nix', 'upgrade-self'];
exports.desc = 'upgrade eask2nix itself';

exports.handler = async (argv) => {
  process.chdir(UTIL.plugin_dir());
  let proc = child_process.spawn('git', ['pull'], { stdio: 'inherit' });

  // You would just need to register the error event, or else it can't print
  // the help instruction below.
  proc.on('error', function () { });

  proc.on('close', function (code) {
    if (code == 0) {
      process.stdout.write('✓ Done upgrading eask2nix to the latest version');
      return;
    }
    // Help instruction here!
    console.log('');
    console.log('');
    console.log('✗ Failed to upgrade eask2nix, possible causes are:');
    console.log('');
    console.log('  [1] Make sure you have git installed and has the right permission');
    console.log('  [2] You install eask2nix with other package management tool');
    console.log('');
    console.log('For example, if you have installed eask2nix with npm:');
    console.log('');
    console.log('    $ npm install -g eask2nix@latest');
    console.log('');
    process.stdout.write('Visit https://github.com/nix-community/eask2nix#installation to see all available install options');
  });
};


cmds/generate.js
/**
 * Copyright (C) 2022 Jen-Chieh Shen
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with GNU Emacs; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

"use strict";

const child_process = require("child_process");

exports.command = ['generate [file..]'];
exports.desc = 'Generate Nix expressions from an Eask-file';
exports.builder = {
  file: {
    description: 'target Eask-file',
    requiresArg: false,
    type: 'string',
  },
};

exports.handler = async (argv) => {
  let okEmacs = await UTIL.checkExec('emacs');
  if (okEmacs != 0) return;

  let okEask = await UTIL.checkExec('eask');
  if (okEask != 0) return;

  // let okNixHash = await UTIL.checkExec('nix-hash');
  // if (okNixHash != 0) return;

  UTIL.e_call(argv, 'generate', argv.file);
};
